### ISSUE TITLE
Feature request: bit-vector array syntax

### ISSUE THREAD

**Issue by seanmcl** [dafny-lang/dafny#1175]
It would be nice to be able to access bit vector fields as booleans.  E.g. 

```
var foo: bv128 = 1839428 as bv128
assert foo[17] == 1 // I'm not sure if this is true
```

**Comment by davidcok** [comment]
How about this syntax:

[1-9][0-9]*#[0-9A-Za-z](_[0-9A-Za-z])*

The part before the # is the radix (as a decimal number). The part after is a sequence of ‘digits’ appropriate to the radix, underscores allowed, case-insensitive, leading 0s allowed, most-significant digits to the left, no embedded white space permitted.

- David

**Comment by RustanLeino** [comment]
These are two distinct proposal, methinks.
@seanmcl is proposing a syntax `foo[17]` that tells you whether or not bit `17` of bitvector `foo` is a `1`.
@davidcok is proposing an extended syntax for numeric literals (in addition to the decimal and hexadecimal forms that are supported today).

**Comment by davidcok** [comment]
Sure. My first thought was that Sean wanted to tell at a glance what a given bit value of a literal is. But the array element syntax is useful here as well

**Comment by robin-aws** [comment]
Would the type of `foo[17]` be `boolean` or `int`? I think we're implying the latter, I just ask because the original comment says "booleans".

**Comment by seanmcl** [comment]
I'd want it to be boolean.

**Comment by RustanLeino** [comment]
I too think it `bool` is good. If you want the result to be a bitvector, you'd still need to write `foo & (1 << 17)`.

I could also see use for the companion operation `foo[n := b]`, which says to set bit `n` if `b` evaluates to `true` to and clear bit `n` otherwise. Again, `b` would be a `bool`.

**Comment by seanmcl** [comment]
Yeah, we have the `foo[n := b]` operator defined in our code, so that would be a nice addition as well.
